meta/templates/sai_adapter.py.tt

```diff
@@ -0,0 +1,487 @@
+        [%- END -%]
+    [%- END -%]
+    [%- # Finally all attributes assigned to the specific operation (like 'get' or 'create') -%]
+    [%- FOREACH attr IN apis.$api.objects.${function.object}.attrs.${function.operation} -%]
+,[% br %][% attr.simple_name %]=None
+    [%- END -%]
+):
+[% END -%]
+
+[%- ######################################################################## -%]
+
+[%- BLOCK declare_variables -%]
+    [%- # Declare variables that are not part of python interface -%]
+    [%- # but are required by thritft functions -%]
+    [%- FOREACH arg IN function.adapter_preprocessed_args -%]
+
+        [%- IF arg.is_list %]
+    [% arg.name %] = []
+        [%- ELSIF arg.is_attr %]
+    [% arg.name %] = None
+        [%- END -%]
+    [%- END -%]
+[%- END -%]
+
+[%- ######################################################################## -%]
+
+[%- ######################################################################## -%]
+
+[%- BLOCK initialize_attribute_from_value -%]
+    [%-  indentation = indentation || 1; tab = 4; indent = ' '; br = "\n     " _ indent.repeat(tab*indentation + variable.length + 21); indent = indent.repeat(tab*indentation) %]
+[% indent %][% variable %]_value = sai_thrift_attribute_value_t([% attr.typename %]=[% attr.simple_name %])
+[% indent %][% variable %] = sai_thrift_attribute_t(id=[% attr.name %],[% br %]value=[% variable %]_value)
+[% END -%]
+
+[%- ######################################################################## -%]
+
+[%- ######################################################################## -%]
+
+[%- BLOCK append_listarg_with_attr %]
+    if [% attr.simple_name %] is not None:
+    [%- PROCESS initialize_attribute_from_value variable = arg.type.subtype.short_name indentation = 2 -%]
+        [% arg.name %].append([% variable %])
+[% END -%]
+
+[%- ######################################################################## -%]
+
+[%- ######################################################################## -%]
+
+[%- BLOCK call_function -%]
+    [%- IF function.operation == 'get' or function.operation == 'stats' -%]
+[% function.rpc_return.name %] = 
+    [%- ELSIF function.rpc_return.type.name != 'void' -%]
+return 
+    [%- END -%]
+client.[% function.thrift_name %](
+[%- comma = 0; FOREACH rpcarg IN function.rpc_args %][% IF comma %], [% ELSE; comma = 1; END %][% rpcarg.name %][% END %])
+[%- END -%]
+
+[%- ######################################################################## -%]
+
+[%- ######################################################################## -%]
+
+[%- BLOCK try -%]
+    try:
+        [%- content -%]
+
+        [%- PROCESS catch_exception %]
+[%- END -%]
+
+[%- ######################################################################## -%]
+
+[%- ######################################################################## -%]
+
+[%- BLOCK catch_exception -%]
+    except sai_thrift_exception as e:
+        status = e.status
```

Comment:
how about generating those exception handlers into a utils method that can be used by the user?

Reply:
Not sure I understand your point of view. This block simply surround each RPC server function call in sai_adapter with try-except clause. How would you like to turn it into a method? Secondly, do you mean to put this into sai_utils file that we want to add here: https://github.com/opencomputeproject/SAI/pull/1335/files#diff-7a25e711c6b68d0df0743a168d6efcdb1258b18e7062942eefd507eec767d9cb or some other? sai_utils file is not intended to be autogenerated. It is written manually.

Reply:
I see, there is a condition 'if CATCH_EXCEPTIONS:'.

Reply:
but still, have some concerns about this try-except, first, it is not thread-safe. secondly, an exception should be happened as unexpected. why do we need to wrap that, and also request a manual check for the status, but not throw it directly and let users to handle them as their needs?

Reply:
This is the design we established and according to that we have written all the tests, this change would mean we have to modify all the test cases we have to handle exceptions.

Reply:
ok, i see